File: benchmarks/allmethod_bench.R
#!/usr/bin/env Rscript
# benchmarks/allmethod_bench.R
# Full-method comparison on stepstone fulldata (1.58M rows, 9 margins).
# Method-specific convergence metrics. All loss values computed post-hoc from weights.

suppressPackageStartupMessages({
  library(arrow); library(jsonlite); library(leafblower)
})
`%||%` <- function(a, b) if (!is.null(a)) a else b

DATA    <- "benchmarks/stepstone_fulldata_bench_data.parquet"
TARGETS <- "benchmarks/stepstone_fulldata_bench_targets.json"
MAX_ITER <- 3000L; MAX_WT <- 5; TOL <- 1e-4
NREPS   <- 3L        # repetitions per method for median timing
# Default convergence per method (tol only; metric set by harvest.R per-method defaults)
DEF <- list(tol=TOL)  # uses each method's natural metric (harvest.R switch)

cat("Loading data...\n")
df_raw <- as.data.frame(read_parquet(DATA))
tgt    <- lapply(fromJSON(TARGETS), function(x){x<-unlist(x);x/sum(x)})
df     <- df_raw[,names(tgt),drop=FALSE]
for(v in names(tgt)) df[[v]] <- factor(df[[v]], levels=names(tgt[[v]]))
n <- nrow(df)
cat(sprintf("n=%s | margins=%d | max_iter=%d | tol=%.0e\n\n",
            format(n,big.mark=","), length(tgt), MAX_ITER, TOL))

STATUS <- c("0"="converged","1"="no_conv","2"="infeasible","3"="bad_arg","4"="budget","5"="stall")

# Post-hoc metric computation from weights
compute_metrics <- function(w, df, tgt) {
  W <- sum(w)
  max_err <- 0; mkl <- 0; wkl <- 0; chi2 <- 0
  w0 <- rep(1, length(w))   # uniform baseline for weight-KL
  for(v in names(tgt)) {
    S <- tapply(w, df[[v]], sum); S <- S / W
    T <- tgt[[v]][names(S)]
    T[is.na(T)] <- 0
    err <- abs(S - T)
    max_err <- max(max_err, max(err, na.rm=TRUE))
    # marginal KL: Σ T*log(T/S)  (skip zero T)
    pos <- T > 0 & S > 0
    mkl <- mkl + sum(T[pos] * log(T[pos] / S[pos]))
    chi2 <- chi2 + sum((S - T)^2 / pmax(T, 1e-12))
  }
  # weight-KL: Σ w_i * log(w_i / w_mean)
  wm <- W / n
  pos <- w > 0
  wkl <- sum(w[pos] * log(w[pos] / wm)) / W
  c(max_err=max_err, marginal_kl=mkl, kl=wkl, chi2=chi2)
}

run <- function(label, conv, ...) {
  times <- numeric(NREPS)
  res_last <- NULL
  for (rep in seq_len(NREPS)) {
    t0 <- proc.time()["elapsed"]
    res <- suppressWarnings(tryCatch(
      harvest(df, tgt, min_weight=0, max_weight=MAX_WT,
              max_iterations=MAX_ITER, convergence=conv, verbose=0L,
              attach_weights=FALSE, ...),
      error=function(e) structure(list(), class="error", message=e$message)
    ))
    times[rep] <- round((proc.time()["elapsed"]-t0)*1000)
    res_last <- res
  }
  if(inherits(res_last,"error")) {
    cat(sprintf("  ERROR: %s\n", attr(res_last,"message")))
    return(NULL)
  }
  ri  <- attr(res_last,"result")
  w   <- res_last  # attach_weights=FALSE: harvest returns numeric weights vector directly
  m   <- compute_metrics(w, df, tgt)
  st  <- STATUS[as.character(ri$status)]
  if(is.na(st)) st <- as.character(ri$status)
  loss_fn <- conv$metric %||% ri$convergence_used$metric %||% "default"
  data.frame(method=label, loss_fn=loss_fn,
             median_ms=median(times), min_ms=min(times), max_ms=max(times),
             status=st, iterations=ri$iterations, algorithm=ri$algorithm,
             max_err=signif(m["max_err"],4), marginal_kl=signif(m["marginal_kl"],4),
             kl=signif(m["kl"],4), chi2=signif(m["chi2"],4),
             stringsAsFactors=FALSE)
}

# Per-method natural default: metric set by harvest.R switch (1ab1165).
# Only tol is set here; each method uses its solver-appropriate objective.
cfg <- list(
  list("oris",                  DEF, method="oris"),
  list("oris + accel",          DEF, method="oris",      accelerate=TRUE),
  list("oris + greedy",         DEF, method="oris",      scheduler="greedy"),
  list("oris + greedy + accel", DEF, method="oris",      scheduler="greedy", accelerate=TRUE),
  list("oris_soft (auto cp)",   DEF, method="oris_soft"),
  list("oris_soft + accel",     DEF, method="oris_soft", accelerate=TRUE),
  list("raking",                 DEF, method="raking"),
  list("greenkhorn",             DEF, method="greenkhorn"),
  list("greenkhorn + greedy",    DEF, method="greenkhorn", scheduler="greedy"),
  list("greenkhorn + accel",     DEF, method="greenkhorn", accelerate=TRUE),
  list("sinkhorn",               DEF, method="sinkhorn"),
  list("chebyshev",              DEF, method="chebyshev"),
  list("greg",                   DEF, method="greg"),
  list("logit",                  DEF, method="logit"),
  list("newton_kl",              DEF, method="newton_kl")
)

results <- vector("list", length(cfg))
for(i in seq_along(cfg)) {
  args <- cfg[[i]]; label <- args[[1]]; conv <- args[[2]]; args <- args[-(1:2)]
  cat(sprintf("[%2d/%2d] %-36s", i, length(cfg), label)); flush.console()
  results[[i]] <- do.call(run, c(list(label=label,conv=conv),args))
  if(!is.null(results[[i]])) {
    r <- results[[i]]
    cat(sprintf("  %7dms [%d-%d]  %-10s  mkl=%.4f  max_err=%.4f  iters=%d\n",
                r$median_ms, r$min_ms, r$max_ms, r$status, r$marginal_kl, r$max_err, r$iterations))
  }
}

tab <- do.call(rbind, Filter(Negate(is.null), results))
tab_s <- tab[order(tab$median_ms),]

cat("\n=== Results (sorted by median_ms) ===\n")
print(tab_s[,c("method","loss_fn","median_ms","min_ms","max_ms","status","iterations",
               "marginal_kl","max_err","kl","chi2","algorithm")],
      row.names=FALSE, digits=4)

outfile <- "benchmarks/results/allmethod_bench_r.csv"
write.csv(tab, outfile, row.names=FALSE)
cat(sprintf("\nSaved: %s\n", outfile))

File: benchmarks/allmethod_bench.py
#!/usr/bin/env python3
# benchmarks/allmethod_bench.py
# Python/R parity benchmark: all methods + accel/greedy on stepstone fulldata.
# Uses improvement rule (stall detection) matching allmethod_bench.R.
# Tier-1 metrics: marginal_kl + wall_time.

import sys, time, json, math, csv, statistics
from pathlib import Path

import pandas as pd
import numpy as np

repo = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(repo / "python"))
from leafblower import harvest

DATA    = repo / "benchmarks/stepstone_fulldata_bench_data.parquet"
TARGETS = repo / "benchmarks/stepstone_fulldata_bench_targets.json"
MAX_IT  = 3000
MAX_WT  = 5.0
TOL     = 1e-4
NREPS   = 3          # repetitions per method for median timing
OUTFILE = repo / "benchmarks/results/allmethod_bench_py.csv"

# Natural convergence per method — improvement rule (stall detection)
KL   = {"metric": "marginal_kl", "rule": "improvement", "tol": TOL}
MERR = {"metric": "max_err",     "rule": "improvement", "tol": TOL}
WKL  = {"metric": "kl",         "rule": "improvement", "tol": TOL}
CHI  = {"metric": "chi2",       "rule": "improvement", "tol": TOL}

_STATUS = {0: "converged", 1: "no_conv", 2: "infeasible", 3: "bad_arg",
           4: "budget",    5: "stall"}

print("Loading data...")
df  = pd.read_parquet(DATA)
with open(TARGETS) as f:
    tgt_raw = json.load(f)
tgt = {k: {lv: v / sum(d.values()) for lv, v in d.items()} for k, d in tgt_raw.items()}
df  = df[[c for c in tgt if c in df.columns]]
print(f"n = {len(df):,} | margins = {len(tgt)} | max_iter = {MAX_IT} | tol = {TOL:.0e}\n")


def post_hoc_mkl(w_arr, df, tgt):
    """Compute obs-level marginal_kl from returned weights via groupby."""
    w = pd.Series(w_arr, index=df.index)
    W = w.sum()
    mkl, max_err = 0.0, 0.0
    for v, targets in tgt.items():
        if v not in df.columns: continue
        s_map = w.groupby(df[v]).sum() / W
        for lv, T in targets.items():
            S = float(s_map.get(lv, 0.0))
            max_err = max(max_err, abs(S - T))
            if T > 0 and S > 0:
                mkl += T * math.log(T / S)
    # weight-KL: sum(w_i * log(w_i / w_mean)) / W
    w_mean = W / len(w_arr)
    wkl = float(np.sum(w_arr[w_arr > 0] * np.log(w_arr[w_arr > 0] / w_mean)) / W)
    return mkl, max_err, wkl


def run(label, conv, **kwargs):
    times = []
    res_last = None
    for _ in range(NREPS):
        t0 = time.perf_counter()
        try:
            res = harvest(df, tgt, min_weight=0.0, max_weight=MAX_WT,
                          max_iterations=MAX_IT, convergence=conv,
                          verbose=0, attach_weights=False, **kwargs)
        except Exception as e:
            elapsed = (time.perf_counter() - t0) * 1000
            print(f"  ERROR: {e}")
            return {"method": label, "loss_fn": conv.get("metric","?"),
                    "median_ms": round(elapsed), "min_ms": round(elapsed), "max_ms": round(elapsed),
                    "status": "error", "iterations": -1, "algorithm": "error",
                    "marginal_kl": float("nan"), "max_err": float("nan"), "kl": float("nan")}
        times.append((time.perf_counter() - t0) * 1000)
        res_last = res
    assert res_last is not None  # exceptions trigger early return; loop always assigns if it completes
    ri   = res_last["result"]
    w_np = np.array(res_last["weights"])
    mkl, merr, wkl = post_hoc_mkl(w_np, df, tgt)
    return {
        "method":      label,
        "loss_fn":     conv.get("metric", "?"),
        "median_ms":   round(statistics.median(times)),
        "min_ms":      round(min(times)),
        "max_ms":      round(max(times)),
        "status":      _STATUS.get(ri["status"], str(ri["status"])),
        "iterations":  ri["iterations"],
        "algorithm":   ri.get("algorithm_name", kwargs.get("method", "?")),
        "marginal_kl": round(mkl, 6),
        "max_err":     round(merr, 6),
        "kl":          round(wkl, 6),
    }


cfg = [
    # oris family — natural: marginal_kl
    ("oris",                  KL,   {"method": "oris"}),
    ("oris + accel",          KL,   {"method": "oris",      "accelerate": True}),
    ("oris + greedy",         KL,   {"method": "oris",      "scheduler": "greedy"}),
    ("oris + greedy+accel",   KL,   {"method": "oris",      "scheduler": "greedy", "accelerate": True}),
    ("oris_soft (auto cp)",   KL,   {"method": "oris_soft"}),
    ("oris_soft + accel",     KL,   {"method": "oris_soft", "accelerate": True}),
    # raking/greenkhorn family — natural: weight-kl (Csiszar-Tusnady monotone)
    ("raking",                 WKL,  {"method": "raking"}),
    ("greenkhorn",             WKL,  {"method": "greenkhorn"}),
    ("greenkhorn + greedy",    WKL,  {"method": "greenkhorn", "scheduler": "greedy"}),
    ("greenkhorn + accel",     WKL,  {"method": "greenkhorn", "accelerate": True}),
    # KL/dual family
    ("sinkhorn",               WKL,  {"method": "sinkhorn"}),
    ("newton_kl",              WKL,  {"method": "newton_kl"}),
    # logit
    ("logit",                  MERR, {"method": "logit"}),
    # chi2 family (R benchmark: chebyshev ~17s, greg ~3s)
    ("chebyshev",              CHI,  {"method": "chebyshev"}),
    ("greg",                   CHI,  {"method": "greg"}),
    ## lbfgsb removed — strictly dominated by newton_kl (O(n) vs O(M_cell) per iter)
]

results = []
for i, (label, conv, kwargs) in enumerate(cfg, 1):
    print(f"[{i:2d}/{len(cfg):2d}] {label:<36}", end="", flush=True)
    r = run(label, conv, **kwargs)
    results.append(r)
    mkl = f"{r['marginal_kl']:.5f}" if math.isfinite(r['marginal_kl']) else "  NaN"
    print(f"  {r['median_ms']:7d}ms [{r['min_ms']}-{r['max_ms']}]  {r['status']:<10}  mkl={mkl}  iters={r['iterations']}")

hdr = ["method","loss_fn","median_ms","min_ms","max_ms","status","iterations","algorithm","marginal_kl","max_err","kl"]
print("\n=== Python Results (sorted by median_ms) ===")
col_w = [36, 12, 10, 8, 8, 11, 10, 14, 13, 10, 8]
fmt = "".join(f"{{:<{w}}}" for w in col_w)
print(fmt.format(*hdr))
print("-" * sum(col_w))
for r in sorted(results, key=lambda x: x["median_ms"]):
    mkl  = f"{r['marginal_kl']:.5f}" if math.isfinite(r.get('marginal_kl', float('nan'))) else "NaN"
    merr = f"{r['max_err']:.5f}"     if math.isfinite(r.get('max_err',     float('nan'))) else "NaN"
    wkl  = f"{r['kl']:.4f}"         if math.isfinite(r.get('kl',          float('nan'))) else "NaN"
    print(fmt.format(r["method"], r["loss_fn"], r["median_ms"], r["min_ms"], r["max_ms"],
                     r["status"], r["iterations"], r["algorithm"], mkl, merr, wkl))

OUTFILE.parent.mkdir(exist_ok=True)
with open(OUTFILE, "w", newline="") as f:
    w = csv.DictWriter(f, fieldnames=hdr)
    w.writeheader()
    w.writerows(results)
print(f"\nSaved: {OUTFILE}")
